docs: add design doc for settings extensions

cbgbt opened the pull request:
**Issue number:**
#2669

**Description of changes:**
This adds a design document detailing the changes that are being made to Bottlerockets' settings system in order to support out-of-tree variant builds. The user interface for settings via `apiclient` should remain the same, but this outlines how developers will interact with the settings system for in-tree and out-of-tree variants.

This document will be moved under `sources/apiserver` as documentation after it is implemented.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

cbgbt committed:
docs: add design doc for settings extensions

cbgbt commented:
I think we should discuss how we'd like to accept pull requests for designs as a project. e.g. I'm partial to how the Rust project organizes RFCs in a [separate repository](https://github.com/rust-lang/rfcs).

Since we don't yet have such a process, I'd like to keep my design docs in the Bottlerocket repository for now. They can be moved later if we determine a better place for them.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,314 @@
+### Setting Storage
+Today, Bottlerocket's `apiserver` utilizes models to perform strict validation of settings inputs, which are then serialized to JSON and stored in Bottlerocket's datastore.
+
+Writes to Bottlerocket settings is going to change in many ways under our new paradigm.
```
nit

```suggestion
Writing to Bottlerocket settings is going to change in many ways under our new paradigm.
```

or

Writes to Bottlerocket settings are going to change in many ways under our new paradigm.

stockholmux commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,314 @@
+These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `---` as such:
+
+```toml
+---
````
I would suggest delimiting front mater with something other than triple dash.

Triple plus (`+++`) is my preference as [Hugo, Forestry, and Zola use it. Some other software uses `---toml`](https://github.com/denoland/deno_std/issues/3094#issuecomment-1400018219) but that is not symmetrical, the ending being triple dash.. yuck. 

Triple dash is likely to confuse people as that alone is widely used for yaml.

webern replied:
I would prefer not to use frontmatter since I'm concerned it will confuse syntax highlighters, linters, etc. Case in point, see the red highlights in this GitHub render. It rather see a separate file with a naming convention, theconfig.meta.toml or theconfig.frontmatter.toml... something like that.

webern commented on docs/design/settings/extensions/settings-extension-example.md:
```diff
@@ -0,0 +1,105 @@
+* Whenever the daemon is reconfigured using the Bottlerocket API, it is automatically restarted with the new configuration.
+
+In order to do this, we need to create the following resources in our Bottlerocket variant:
+* A setting extension called `mydaemon` to add the new value `settings.mydaemon.query-url` to the API.
```
Should it be called mydaemon-settings to distinguish it from the mydaemon daemon binary?

```suggestion
* A setting extension called `mydaemon-settings` to add the new value `settings.mydaemon.query-url` to the API.
```

webern commented on docs/design/settings/extensions/settings-extension-example.md:
````diff
@@ -0,0 +1,105 @@
+
+```
+├── packages
+│   └── hello-agent
````
```suggestion
│   └── mydaemon
```

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,314 @@
+Let us cast this requirement in a different light: imagine an OS administrator wishes to restrict what values can be written to an existing setting -- for example, they may wish to only allow settings values to `network.hosts` so long as they include as a subset the entries mandated by their organization.
+To allow this, settings extensions will be capable of register themselves as a *validator* for any other setting on the host, including settings that they do not own.
+When handling write requests for multiple settings in a single transaction, the `apiserver` will first gather the resulting writes that would occur under each settings extension involved in the writes.
+Afterwards, the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.
```
```suggestion
Then the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.
```

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,314 @@
+```toml
+# auditor.toml -- setting extension config file for `settings.auditor`
+[[extension.validations]]
+validates = "network.hostname"
````
```suggestion
validates = [ "network.hostname" ]
```

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,314 @@
+[[extension.validations]]
+validates = "network.hostname"
+with-templates = [
+    "/templates/auditor-extension/validator-template.toml"
```
I'm not sure I understand how a template is involved in the validation.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,314 @@
+
+Any setting can be written or read at any of the `supported-versions` listed by the extension.
+Unversioned requests to read or write this setting are assumed to be the `default-version`.
+During a `set` transaction, the `apiserver` will consider the provided version to be the `canonical-version`, and will ask the settings extension to perform migrations from the `canonical` version to all supported versions using the `migrate` command.
```
Should `migrate` be rolled into `set`? i.e. should `set` also do the "migrations" since there is no case when `set` would *not* be followed by `migrate`?

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,314 @@
+$ schnauzer --required-extension 'hostname=v1' '{{ network.hostname }}'
+```
+
+As an optimization, templates which render to `systemd` units will invoke a restart command to systemd
````
Some might not be running yet during the boot sequence.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,314 @@
+
+```bash
+# hostname.toml setting extension config file
+[templating.v1] # template helpers are versioned
+helpers = [
+    "localhost_aliases"
+]
````
Something seems off here. It looks like there would be multiple versions of the schema by which you can specify which helpers exist.

Maybe this? I guess I'm not super clear on the versioning of helpers.

```suggestion
[templating.helpers] # template helpers are versioned
localhost_aliases = "v1"
```

cbgbt replied:
This is by design.

Each existing version of a setting would have its own helpers that it provides. During template rendering, we request a settings at a given version, we need to ensure that the template helpers at that version continue to exist and provide the same API.

In practice, this likely means that you'd specify more or less the same helpers at all versions, but this gives us a valve to move to new versions.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,314 @@
+Bottlerocket settings currently have the ability to compute a default value via something called a "settings generator".
+During boot, settings extensions are given a chance to populate settings with default values that are then persisted, effectively determining the setting value for the lifetime of that instance.
+
+Settings generation often depends on the presence of other settings; therefore the invocation protocol will allow settings generation to accept rendered templates as input:
```
rendered templates instead of other settings? wouldn't be more convenient for me to say I need `aws.region` than to find a template which contains that value?

cbgbt replied:
This design made `aws.region` more opaque from the apiserver's perspective, so it wouldn't be possible for the server to resolve that. It's one of the reasons why we'd need to move system services to use template files. Basically anything that would request a setting would do so via a template.

cbgbt commented:
After prototyping and a few virtual discussions about this design, we've opted to move to something a bit simpler, opting instead to be more JSON-based at the expense of extension flexibility regarding storage and formatting.

I'll move this to draft until I have the time to update the document to reflect the current direction. For now I'm focusing on fleshing out the prototype a bit more to make sure I catch any other lurking gotchas.

cbgbt commented:
After spending time [prototyping the Settings SDK](https://github.com/cbgbt/bottlerocket-settings-sdk/tree/0bee8537a08cb015fbaa7ebd769fe98d0d72aaba), considering some of the feedback here, and thinking about how to simplify the settings design, I've significantly reworked this doc to account for my revised thinking.

Brief Changelog:
* Added more explicit requirements and exposition to help better-set the stage for the purpose of these changes.
* Removed RPM-packaging specifications to be discussed in greater depth in a separate document. This is a complex problem, and this doc is already a bit too long.
* Reworked the interfaces for settings extensions to strictly involve passing JSON messages.
* These JSON changes now have the serialized settings data owned entirely by the OS, even if the objects are opaque to it.

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+Command line arguments are used to signify the "request" being made of the extension.
+Exit codes will be used to signal the status of the request, with formatted output being delivered on `stdout`, and logs delivered on `stderr`.
+
+As an example, suppose a user wants to set the "message of the day" for their instance:
+```bash
+$ apiclient set motd="Hello, Bottlerocket!"
+```
````
How will the type inference work? Same sort of approach as [massage_set_input](https://github.com/bottlerocket-os/bottlerocket/blob/d5acb9e7b33121f0bab1f3102b2b8c295d750ed0/sources/api/apiclient/src/main.rs#L588) today?

cbgbt replied:
Hmm, I hadn't considered this. I think we could resolve this via the same algorithm with the same caveats. It's worth considering if we can come up with a cleaner solution here, but I don't think failing to come up with one introduces a change in behavior.

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+Then the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.
+Any one settings extension has the power to halt the transaction by returning a non-zero exit-code on validation.
+
+Validations must specify the setting which triggers the validation in their configuration file:
```
nit: is this reading correct?
```suggestion
Extensions must specify the setting which triggers the validation in their configuration file:
```

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+└── lib
+    └── bottlerocket
+        └── settings
+            ├── config.d
+            │   └── network.toml
+            └── extensions.d
+                └── network -> /usr/bin/network-extension
```
Is the 1:1 match between config file and symlink name enforced, or just a convention?

cbgbt replied:
It's enforced: This is how the system locates the configuration file for the given extension.

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+
+#### Transactional Writes and Cross-Validation
+Bottlerocket today is missing a mechanism for cross-validation of settings.
+As an example, suppose we want to represent a range of integers in settings by allowing a `setting.min-value` and `setting.max-value` to be set -- it should be possible for settings extensions to set both simultaneously, while also verifying that the state from setting both is valid (`setting.min_value <= setting.max_value`).
```
Does the switch from `min-value` to `min_value` signify anything?

cbgbt replied:
Whoops! That's a mistake.

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+
+```toml
+# auditor.toml -- setting extension config file for `settings.auditor`
+[extension.validates]
+network = "v1"
````
Are extensions allowed to validate top-level settings that don't exist? E.g.,
```
[extension.validates]
not-found = "v1"
```

What about versions that the extension doesn't support?

cbgbt replied:
Good questions. I'll update the doc, but the answer to both is no.

At the spec/packaging level, you'd need to specify a dependency on settings that are required. We can add a build-time check to ensure that validated versions are supported.

At runtime either of these will result in an error occurring and the setting failing to be set.

cbgbt replied:
I've changed my mind about this. I think the first case is still valid. If you attempt to validate a setting which isn't on the system, then that will silently succeed.

This could happen in cases where a vendor creates a settings extension that you are interested in, and it takes on many responsibilities but you don't particularly care for the validation functionality.

A version not being present will cause validation to fail. In this case, the author clearly wanted a validation to occur against settings which are present on the system but it cannot succeed.

Settings should have a `Requires` line in their RPM package specifying their related settings to prevent these issues at build time.

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+```
+
+Any setting can be written or read at any of the `supported-versions` listed by the extension.
+Unversioned requests to read or write this setting are assumed to be the `default-version` and should only take place when a user makes a request using the CLI -- all automated process *should* refer to an explicit version.
````
Unversioned requests will also be used when handling user-data, right?

cbgbt replied:
Yes!

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+Unversioned requests to read or write this setting are assumed to be the `default-version` and should only take place when a user makes a request using the CLI -- all automated process *should* refer to an explicit version.
+During a `set` transaction, the `apiserver` will consider the provided version to be the `canonical-version`, and will ask the settings extension to perform migrations from the `canonical` version to all supported versions using the `migrate` command.
+These migrations always begin at the `canonical` version, and settings extensions are free to perform the migrations however they wish; however, Bottlerocket's Settings SDK will provide utilities to help settings developers support a series of linear setting upgrades and downgrades.
+Read on for more information about the Settings SDK
```
```suggestion
Read on for more information about the Settings SDK.
```

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+Read on for more information about the Settings SDK
+
+All migrations are performed and stored at setting-write-time because Bottlerocket rollbacks may result in running an OS with a settings extension which does not support the version at which a given setting is stored.
+If Bottlerocket boots and finds that a setting is stored at a value which is not supported by it's installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent dataloss in the case that the image can eventually update to a new image which supports the stored setting version.
```
```suggestion
If Bottlerocket boots and finds that a setting is stored at a value which is not supported by its installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent data loss in the case that the image can eventually update to a new image which supports the stored setting version.
```

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+With the introduction of settings extensions, this will change such that templates specify `(name, version)` tuples for all extensions needed to render the template.
+These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `+++` as such:
+
+```toml
++++
+[required-extensions]
+hostname = v1
+kubernetes = v1
++++
````
I still have some doubts about the "front matter" approach, but I'm willing to defer judgment until I see the implementation 😀 

I'm still somewhat partial to the filesystem-oriented approach, e.g.:
```
usr/
└── share
    └── templates
        └── etc
            └── kubelet
                ├── kubelet.config
                └── kubelet.config.required.d
                    ├── kubernetes
                    └── network
```

Since this is a bit more consistent with some of the other ways dependencies are modeled, at the cost of giving up tight binding to the template.

webern replied:
My take was something like this, where the metadata file is found by naming convention:

```
usr/
└── share
    └── templates
        └── etc
            └── kubelet
                ├── kubelet.config
                └── kubelet.config.info.toml
```

cbgbt replied:
I'll open another PR discussing particulars about this design.

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+```
+
+This will allow Bottlerocket to find the owning extensions for the data needed to render the template.
+It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "usptream" changes.
````
```suggestion
It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "upstream" changes.
```

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+
+## Settings Extension Development and the Settings SDK
+Settings extensions can be written in any programming language, so long as a symlink to a binary speaking the "extension protocol" lands in the right spot on the filesystem.
+That said, many settings extensions need to tackle the same challenges: How do we represent and implement settings versions? Migrations? Validation? Template utilities? Rather than forcing all extensions to implement this in isolation, we will create a shared library, called the "Settings SDK", which will be built in Bottlerocket's monorepo and published to crates.io.
```
Let's not name this or say what repo it will be built from in this doc; it's not clear that we'd put this crate in the core repo vs. having another dedicated project for it.

I would like to avoid having a wide variety of different artifacts published from a single repo, since that complicates CI/CD workflows and the releases page.

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+Indeed, even Bottlerocket core developers wish for tooling to better generate and validate EC2 userdata without the need to interact with the EC2 APIs.
+
+This feature is possible even in a world where settings are defined by extensions which are discovered dynamically.
+During a variant build, we can modify sodapop to extract the settings extension binaries and construct a Docker container which can use the information exposed by those binaries in order to validate userdata, or render a userdata template.
```
```suggestion
During a variant build, we can modify twoliter to extract the settings extension binaries and construct a Docker container which can use the information exposed by those binaries in order to validate userdata, or render a userdata template.
```

bcressey commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+### How will the `apiserver` handle concurrent requests in the face of transactional writes?
+The `apiserver` will use a [filesystem-based](https://man7.org/linux/man-pages/man2/flock.2.html) reader-writer lock to ensure that the datastore is only accessed by concurrent readers or a single writer. Writes which are sent to the `apiserver` while the file is lock are placed in a queue.
+
+## Appendices
+### A. Assisted EC2 Userdata Generation and Validation
```
Would it be better to tackle this in a follow-up design? I agree that it's a potential benefit, but it seems orthogonal to the main concepts here.

cbgbt replied:
Totally fair. It's included here to capture some verbal discussion of this document that occurred and I wanted to capture it as a possible feature of the new system, but it can be delegated to another doc.

bcressey commented on docs/design/settings/extensions/settings-extension-example.md:
````diff
@@ -0,0 +1,112 @@
+```rust
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq, Clone)]
+struct MyDaemonSettingV1 {
+    query-url: Url,
````
```suggestion
    query_url: Url,
```

bcressey commented on docs/design/settings/extensions/settings-extension-example.md:
````diff
@@ -0,0 +1,112 @@
+```
+The configuration template includes "frontmatter" which informs the settings system to use settings values owned by the `mydaemon` extension, then the rest of the template is used to render the configuration file.
+
+## Service File
+The settings system needs to know how to restart `mydaemon` so that it can parse changes to its config file after settings have been changed.
+In order to do that,you will create a `mydaemon-service.toml` file.
+This file will be referred to via a symlink on the Bottlerocket filesystem which instructs the settings system to restart the `mydaemon-service` when changes to `settings.mydaemon` take place:
````
I find it hard to follow these files without a picture of the final install layout - the examples here tell me how to arrange the source files, but not what the destination layout should be, so I can't visualize the resulting spec file.

cbgbt replied:
I've given an example of how files will be arranged on the filesystem -- let me know if that seems ok to you.

zmrow commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+As an example, to use the `localhost_aliases` helper above from a template, your template may look something like this:
+
+```
+{{ hostname.localhost_aliases(arg1, arg2) }}
````
Perhaps I missed it, but how are the helpers registered with handlebars?

cbgbt replied:
The algorithm is something like:

* The template rendering system will use the frontmatter section of the template to determine which settings extensions it needs to use to render the document.
* The template system uses the API to determine what handlebar helpers are provided by the settings extensions required.
* For each helper discovered, the template renderer will register a helper where the implementation results in an API call to execute the settings extension's method for calling that helper.

zmrow commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+```
+
+### Settings Generation
+Bottlerocket settings currently have the ability to compute a default value via something called a "settings generator".
````
I wouldn't say that computing a default value is settings generator's main purpose.  Their main purpose is to calculate settings at runtime for things that can't be given a static default.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+$ apiclient set motd="Hello, Bottlerocket!"
+```
+
+The apiserver would load the current value of `motd` and then create a new object based on the inputted string.
````
I may have already asked this, but why does it need to load the current value before proceeding with the rest of this?

cbgbt replied:
You may be remembering an earlier form of this design where `get` requests would be routed to the extension to respond to. The idea here is that the API server can handle any arbitrary `get` request by referring to the stored JSON document.

Since the API server is also completely capable of handling arbitrary patches to JSON objects, it performs sets by changing the values requested by the user, and then just asking the settings extensions to validate them for correctness.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+```toml
+# auditor.toml -- setting extension config file for `settings.auditor`
+[extension.validates]
+network = "v1"
````
Would this also be possible (i.e. validating only a subsection of a top-level setting?)

```toml
# auditor.toml -- setting extension config file for `settings.auditor`
[extension.validates]
network.foo.bar = "v1"
```

cbgbt replied:
With this design it's not possible. The setting extension itself would have to determine whether or not relevant subtrees of other settings are changed.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
```diff
@@ -0,0 +1,256 @@
+This will allow Bottlerocket to find the owning extensions for the data needed to render the template.
+It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "usptream" changes.
+
+`schnauzer` can also continue to be invoked as a setting generator without the need for a templated configuration file.
```
Is this considered a temporary arrangement?

cbgbt replied:
I think this is actually just wrong, settings generators in concept go away. What may be relevant is that you could invoke schnauzer to arbitrarily render a view of Bottlerocket settings without needing to invoke the API.

webern commented on docs/design/settings/extensions/modular-settings-extensions.md:
````diff
@@ -0,0 +1,256 @@
+```
+
+### Extension Naming & Setting Ownership
+Every top-level setting in Bottlerocket's API must be owned by a singular extension, and that extension will share a name with the setting.
````
I think I might want to write a single extension that handles multiple settings. Will that be possible?

cbgbt replied:
Can you given an example of what you'd want here? It's not possible with this design, though you could probably get pretty close to what you want by sharing code via shared rust crates and taking advantage of cross-validation.

webern commented on docs/design/settings/extensions/settings-extension-example.md:
```diff
@@ -0,0 +1,112 @@
+│   └── mydaemon
+│       ├── build.rs
+│       ├── Cargo.toml
+│       ├── mydaemon.spec
```
There's enough boilerplate including main.rs etc that I'm wondering if in the future sodapop or the internal build system could generate, at least the package.

webern reviewed (approved):
Nice work.

cbgbt commented:
^ Addresses feedback from reviewers.

bcressey commented on docs/design/settings/extensions/settings-extension-example.md:
````diff
@@ -0,0 +1,156 @@
+
+For example, the extension for `my-setting` may be installed like so:
+```
+sys-root
+└── usr
+    ├── libexec
+    │   ├── my-setting-binary
+    │   └── settings
+    │       └── my-setting -> ../my-setting-binary
+    └── share
+        └── settings
+            └── my-setting.toml
````
nit: It'd help to stick with the `mydaemon` example to show how it all fits together.

```suggestion
sys-root
└── usr
    ├── libexec
    │   ├── mydaemon-settings
    │   └── settings
    │       └── mydaemon-> ../mydaemon-settings
    └── share
        └── settings
            └── mydaemon.toml
```

bcressey commented on docs/design/settings/extensions/settings-extension-example.md:
````diff
@@ -0,0 +1,156 @@
+                    └── motd.conf
+```
+
+`motd.template.rendered-to/motd.conf` contains instructions similar to `systemd-tmpfiles.d` explaining locations to which the rendered template will be placed.
+The format of this file dictates the resulting file location and mode of the resulting file:
````
Clever, I missed the implications of this before. I like reusing part of the `tmpfiles.d` syntax for this. It might be good to also support setting UID / GID in addition to mode.
```
/etc/mydaemon/mydaemon.json - - -
/etc/mydaemon.env 0755 root root
```

bcressey reviewed (approved):
Nice work.

cbgbt commented:
^ Addressed feedback from @bcressey.

stockholmux commented:
@cbgbt: I know this is long-closed and merged, but re-reading a few months on, I have some follow up questions:
- would existing settings be implemented in this way as well (e.g. `settings.kubernetes.*`, `settings.network.*`, etc.)?
- is there any changes anticipated from the existing settings?
